pkg/webhook/core.oam.dev/v1alpha2/application/mutating_handler.go

```diff
@@ -0,0 +1,79 @@
+/*
+Copyright 2022 The KubeVela Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package application
+
+import (
+	"context"
+	"encoding/json"
+	"net/http"
+
+	utilfeature "k8s.io/apiserver/pkg/util/feature"
+	"k8s.io/utils/strings/slices"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/webhook"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+
+	"github.com/oam-dev/kubevela/apis/core.oam.dev/common"
+	"github.com/oam-dev/kubevela/apis/core.oam.dev/v1beta1"
+	"github.com/oam-dev/kubevela/pkg/auth"
+	"github.com/oam-dev/kubevela/pkg/features"
+)
+
+// MutatingHandler adding user info to application annotations
+type MutatingHandler struct {
+	Decoder *admission.Decoder
+}
+
+var _ admission.Handler = &MutatingHandler{}
+
+// Handle mutate application
+func (h *MutatingHandler) Handle(ctx context.Context, req admission.Request) admission.Response {
+	if !utilfeature.DefaultMutableFeatureGate.Enabled(features.AuthenticateApplication) {
+		return admission.Patched("")
+	}
+
+	if slices.Contains(req.UserInfo.Groups, common.Group) {
+		return admission.Patched("")
+	}
+
+	app := &v1beta1.Application{}
+	if err := h.Decoder.Decode(req, app); err != nil {
+		return admission.Errored(http.StatusBadRequest, err)
+	}
+
+	auth.SetUserInfoInAnnotation(&app.ObjectMeta, req.UserInfo)
```

Comment:
Q. So if the feature gate enabled, does it always overwrite the user info annotation with the last-modified user?

Reply:
Yes, it will. Another feature is introduced now. By default, the User will not be injected into Application. Only Groups will specified pattern will be injected. If --authentication-with-user flag is used in controller bootstrap parameter, it will do so.

Reply:
BTW, even if this feature gate is not enabled, the previous ServiceAccount authentication will still work as well. The major changes made by this feature gate is:
1. Use the MutatingWebhook to inject user info into Application.
2. Dispatching resources will always send requests with impersonation in avoid of using the controller's identity. This is done by automatically adding Impersonate-User if user not exists in application.
So when the feature gate is not enabled, if the Application carries ServiceAccount annotation, the Impersonate-User will be set to the SA and not be overriden by the default user.